### ISSUE TITLE
Custom dataset labels change if numpy dtype is uint8

### ISSUE THREAD

**Issue by edxsx** [pytorch/pytorch#53959]
## 🐛 Bug

Loading `uint8` numpy arrays that contain labels causes the dataset object to change them into an unknown decimal number. Adding `.astype('int32')` to the end of np.load fixes the issue. No clue why its happening but it completely messes up the training as the optimizer will train the model to return 0.

## To Reproduce

1. Create custom dataset with one-hot labels stored in .npy files with type `uint8`. Applying a normalization transform to the images will also change the labels into an unknown decimal number.
2. Seems like adding `.astype('int32')` to the np.load will fix the issue.

```
class HuBMAPDataset(Dataset):
    def __init__(self, image_path='train_image_tiles/', mask_path='train_mask_tiles/', 
                 tfm=None, mean=None, std=None):
        super().__init__()
        self.image_path = image_path
        self.mask_path = mask_path
        self.nrm = transforms.Normalize(mean=mean, std=std)
        self.tfm = tfm
        self.image_list = os.listdir(self.image_path)

        assert len(self.image_list) == len(os.listdir(self.mask_path))
        self.mask_list = [x[:-5]+'.npy' for x in self.image_list]


    def __getitem__(self, idx):
        image = io.imread(self.image_path + self.image_list[idx])
        # adding .astype('int32') to the next file fixes the issue!!!!!!!!!!!
        mask = np.load(self.mask_path + self.mask_list[idx])
        print(mask.dtype)

        seed = np.random.randint(81261917)
        torch.manual_seed(seed) # needed for torchvision
        image = self.tfm(image)
        image = self.nrm(image)

        torch.manual_seed(seed) # needed for torchvision
        mask = self.tfm(mask)
    
        return image, mask

    def __len__(self):
        return len(self.image_list)


mean = [0.6214918394883474, 0.522968004147211, 0.6425345639387766]
std = [0.2847020526727040, 0.312313973903656, 0.2859026342630386]

tfm = transforms.Compose([transforms.ToTensor(), 
                          transforms.RandomRotation(90, expand=False, center=None, fill=None), 
                          transforms.RandomVerticalFlip(p=0.5), 
                          transforms.RandomHorizontalFlip(p=0.5)])

dataset = HuBMAPDataset('train_image_tiles/', 'train_mask_tiles/', tfm=tfm, mean=mean, std=std)
n_samples = len(dataset)
batch_size = 4
n_val = int(0.1*n_samples)
n_train = n_samples - n_val
n_workers = 0
n_channels = 3

train, val = random_split(dataset, [n_train, n_val], generator=torch.Generator().manual_seed(42))
train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True, 
                          num_workers=n_workers, pin_memory=False, drop_last=True)
val_loader = DataLoader(val, batch_size=batch_size, shuffle=False, 
                        num_workers=n_workers, pin_memory=False, drop_last=True)


```

## Expected behavior

I expect the labels to be integers. But instead 1's are replaced with 0.0039

```
tensor([0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
```

## Environment
PyTorch version: 1.8.0+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.9.0
[pip3] torchvision==0.9.0+cu101
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @fmassa @vfdev-5

**Comment by vfdev-5** [comment]
@edxsx you should not apply `ToTensor` on mask: `mask = self.tfm(mask)` as it also scales the data. Please, see the note here: https://pytorch.org/vision/stable/transforms.html#torchvision.transforms.ToTensor
